lib/stat_tracker.rb

```diff
@@ -14,34 +21,202 @@ def self.from_csv(data_files)
   end
 
   def game_collection
-    GameCollection.new
+    GameCollection.new(@games)
   end
 
   def team_collection
-    TeamCollection.new
+    TeamCollection.new(@teams)
   end
 
   def game_team_collection
-    GameTeamCollection.new
+    GameTeamCollection.new(@game_teams)
+  end
+
+  def game_collection_to_use
+    game_coll = instantiate_game_collection
+  end
+
+  def instantiate_game_collection
+    @instantiate_game_collection ||= game_collection.all
+  end
+
+  def game_team_collection_to_use
+    instantiate_game_team_collection
+  end
+
+  def instantiate_game_team_collection
+    @instantiate_game_team_collection ||= game_team_collection.all
+  end
+
+  def team_collection_to_use
+    instantiate_team_collection
+  end
+
+  def instantiate_team_collection
+    @instantiate_team_collection ||= team_collection.all
   end
 
   def highest_total_score
-    top_score = 0
-    CSV.foreach(@games, headers: true, header_converters: :symbol) do |game|
-      if game[:away_goals].to_i + game[:home_goals].to_i > top_score
-        top_score = game[:away_goals].to_i + game[:home_goals].to_i
-      end
+    total = game_collection_to_use.max_by do |game|
+      game.away_goals.to_i + game.home_goals.to_i
     end
-    top_score
+
+    total.home_goals.to_i + total.away_goals.to_i
   end
 
   def lowest_total_score
-    lowest_score = 1_000_000_000
-    CSV.foreach(@games, headers: true, header_converters: :symbol) do |game|
-      if game[:away_goals].to_i + game[:home_goals].to_i < lowest_score
-        lowest_score = game[:away_goals].to_i + game[:home_goals].to_i
+    total = game_collection_to_use.min_by do |game|
+      game.away_goals.to_i + game.home_goals.to_i
+    end
+
+    total.home_goals.to_i + total.away_goals.to_i
+  end
+
+  #################START of SEASON STATS######################
+
+  def team_name_based_off_of_team_id(team_id)
+    team_collection_to_use.each do |team|
+      return team.team_name if team_id == team.team_id
+    end
+  end
+
+  def games_by_season(season_id)
+   game_collection_to_use.select do |game|
+     season_id == game.season
+   end
+  end
+
+  def season_games_grouped_by_team_id(season_id)
```

Comment:
Returns hash of game objects with keys of game id and values of game objects that match the season